test/uk/gov/hmrc/apprenticeshiplevy/controllers/auth/AuthActionSpec.scala

```diff
@@ -0,0 +1,209 @@
+        .thenReturn(())
+
+      val authAction = new PrivilegedAuthActionImpl(mockAuthConnector)
+      val controller = new Harness(authAction)
+
+      val result = controller.onPageLoad()(FakeRequest())
+      status(result) shouldBe OK
+      contentAsString(result) should include("None found")
+    }
+  }
+
+  "AllProviderAuthAction" must {
+    "return an instance of auth action" in {
+      new AllProviderAuthActionImpl(mockAuthConnector).apply(EmploymentReference(""))
+        .isInstanceOf[AuthAction] shouldBe true
+    }
+
+    "authenticate a privileged application" in {
+      when(mockAuthConnector.authorise[Enrolments ~ LegacyCredentials](any(), any())(any(), any()))
+        .thenReturn(paRetrieval)
+
+      val authAction = new AllProviderAuthActionImpl(mockAuthConnector).apply(EmploymentReference(""))
+      val controller = new Harness(authAction)
+
+      val result = controller.onPageLoad()(FakeRequest())
+      status(result) shouldBe OK
+      contentAsString(result) should include("None found")
+    }
+
+    "authenticate an IR-PAYE enrolled user" when {
+      "the request emp ref matches their emp ref" in {
+        when(mockAuthConnector.authorise[Enrolments ~ LegacyCredentials](any(), any())(any(), any()))
+          .thenReturn(ggRetrieval)
+
+        val authAction = new AllProviderAuthActionImpl(mockAuthConnector).apply(EmploymentReference("123/ABCDEF"))
+        val controller = new Harness(authAction)
+
+        val result = controller.onPageLoad()(FakeRequest())
+        status(result) shouldBe OK
+        contentAsString(result) should include("123/ABCDEF")
+      }
+    }
+
+    "return unauthorized for an IR-PAYE enrolled user" when {
+      "the request emp ref does not match their emp ref" in {
+        when(mockAuthConnector.authorise[Enrolments ~ LegacyCredentials](any(), any())(any(), any()))
+          .thenReturn(ggRetrieval)
+
+        val authAction = new AllProviderAuthActionImpl(mockAuthConnector).apply(EmploymentReference("123%2FABCDE"))
+        val controller = new Harness(authAction)
+
+        val result = controller.onPageLoad()(FakeRequest())
+        status(result) shouldBe UNAUTHORIZED
+      }
+
+      "their is no emp ref returned" in {
+        when(mockAuthConnector.authorise[Enrolments ~ LegacyCredentials](any(), any())(any(), any()))
+          .thenReturn(emptyGGRetrieval)
+
+        val authAction = new AllProviderAuthActionImpl(mockAuthConnector).apply(EmploymentReference("123/ABCDEF"))
+        val controller = new Harness(authAction)
+
+        val result = controller.onPageLoad()(FakeRequest())
+        status(result) shouldBe UNAUTHORIZED
+      }
+    }
+  }
+
+  val enrolments: Enrolments = Enrolments(Set(
+    Enrolment("IR-PAYE", Seq(
+      EnrolmentIdentifier("TaxOfficeNumber", "123"),
+      EnrolmentIdentifier("TaxOfficeReference", "ABCDEF")),
+      "")
+  ))
+  val paRetrieval: Enrolments ~ LegacyCredentials =
+    new ~(Enrolments(Set()), PAClientId("app-id"))
```

Comment:
Something like this might neaten things up?


```
object AuthActionSpec {
  implicit class RetrievalsUtil[A](val retrieval: A) extends AnyVal {
    def `~`[B](anotherRetrieval: B): A ~ B =  uk.gov.hmrc.auth.core.retrieve.~(retrieval, anotherRetrieval)
  }
}
```

Reply:
Wait just seen theres a retrievalOps trait... why arent we using it?